### ISSUE TITLE
Action on release?

### ISSUE THREAD

**Issue by LinuxJoeUser** [kozec/sc-controller#290]
I like to keep my sprint command (usually the LSHIFT key) on LPAD click and for it to stay on until I lift my thumb from the pad.  This is the default behaviour in the _Borderlands_ games, for instance.  In games like _5089_ and _Signal from Tolva_ , where the LSHIFT key must be held down, I've figured out a way to set it up on Steam that makes use of the "Release" activator, and I'd love to figure out a way to do it on SCC.

For now I've set up LPAD to initiate sprint on click and keep it on until I press it again via `cycle(press(button(Keys.KEY_LEFTSHIFT)), release(button(Keys.KEY_LEFTSHIFT)))`, and it's fine, but it really irks me to think that Steam can let me do something with my controller that SCC can't. Lol.  Would there be a non-world-getting-set-on-fire way to create such a thing?  

I'm not sure what form it might take, maybe something like the opposite of the `click()` modifier, an _`unclick()`_ modifier?  That fires an action when a pressed/touched button/pad is released?   Or an _`untouch()`_ modifier to be used on pads only?  Or an _"L/R Pad Release"_ modeshift "Press" option?

I don't know what is or isn't possible, but I do know that if it is possible, kovek is the guy who can find a way! :)

**Comment by kozec** [comment]
> kovek is the guy who can find a way! :)

Who? :D

Just to be clear, are you trying to get shift pressed when pad is pressed down, or just while it is being touched?

**Comment by LinuxJoeUser** [comment]
> Just to be clear, are you trying to get shift pressed when pad is pressed down, or just while it is being touched?

Both, actually!   But let's put my specific case aside for just now because I think it will only confuse things.

The easiest way to explain it is to say we need the same thing as the "Release button" activator in the Steam config tool.  

The action would only occur when the input *ceased* being pressed/touched.

As with Steam's "Release button" activator, we could have this new `repeat(action)` modifier be one of the modifiers available in the Mode Shift menu:  *Press, Hold, Double-click,* and *Release.*  

Admittedly, having a `release(button) action, and a `release(action)` modifier might make things confusing, so a better name would be a plus.  

I'm playing around with various custom actions to see if it can already be done in SCC, but so far no luck.

**Comment by kozec** [comment]
Ok, that shouldn't be hard, although it really looks like those names are going to be inconsistent... I'll write renaming click, doubleclick to clicked, doubleclicked and _something_ into todo.

**Comment by kozec** [comment]
Ok, so now there is

 - click              - activates only if button is being pressed at same time, holds
 - pressed         - activates _when_ button is pressed, for short moment
 - released        - activates when button is released, for short moment
 - doubleclick   - activates if button is double-clicked, holds
 - hold              - activates after button is pressed for a while, holds

Plus, with those new, one can create this kind of madness...

```
released(press(button(Keys.KEY_A))) and pressed(release(button(Keys.KEY_A)))
```

Which basically maps key "A" to reversed state of button. It's pressed if button is not.

**Comment by LinuxJoeUser** [comment]
> I'll need linguist for this stuff...

LOL!  I know, right?  This is why it took me three days to figure out how to even say what I wanted!  Beautiful solution, btw!

**Comment by LinuxJoeUser** [comment]
Msr. Kozec, what's the status of this issue nowadays?  I don't see it if it has.  Am I missing something?  Just wondering.  :)

**Comment by kozec** [comment]
It should work, with `untouched` doing that shift thing you asked for originally.

**Comment by LinuxJoeUser** [comment]
Hmmm.  So far `pressed()`, `released()`, `touched()`, and `untouched()` all result in the error: `Unknown action`\`released\`

Your example above: 

    released(press(button(Keys.KEY_A))) and pressed(release(button(Keys.KEY_A)))

gets the same error, but 

    release(press(button(Keys.KEY_A))) and press(release(button(Keys.KEY_A)))

*without* the "-ed"s gets no errors; (though, I'm not sure what it would do ;) )

Lastly, I had hoped to see *"Left Pad Untouched"* and *"Right Pad Untouched"* added to the list of Mode modifiers in the GUI, right after  *"Left Pad Touched"* and *"Right Pad Touched"*

(Also, your Wiki update has two entries for "Pressed(action)" and "Released(action)".  I think the second set is supposed to be for "Touched(action)" etc.)

**Comment by kozec** [comment]
Just to check, what version are you using? Because that 1st one definitely should work.

**Comment by LinuxJoeUser** [comment]
Whichever version was available via `git pull` two days ago.  I  just updated again this morning and in "About"  it says v3.16.  Is there a way to get a more precise version number if it were available, like 3.16.xx, etc.?

Anyway, maybe I'm doing something wrong here?  Here's what I'm seeing:

![sccuntouched](https://user-images.githubusercontent.com/23580979/34723272-fdb8d082-f517-11e7-8641-6daef1dc4a36.png)

How else should I test it?

**Comment by kozec** [comment]
> Whichever version was available via git pull two days ago. I just updated again this morning and in "About" it says v3.16. Is there a way to get a more precise version number if it were available, like 3.16.xx, etc.?

Sounds like something broke, it should say 0.4.x. What exactly gives you `git pull`? Aren't you on testing branch?

**Comment by LinuxJoeUser** [comment]
I feel like such an idiot.  Fuck me.  No, I wasn't.  I was using Master.   Ok, testing cloned ... but it's giving me

    fatal error: Python.h: No such file or directory
     #include <Python.h>

when I try and build ... pretty sure I have all the dependencies ... any suggestions?  If not, no probs, I'll add a launcher pointing to /run.sh.

**Comment by kozec** [comment]
> fatal error: Python.h: No such file or directory

Check if `/usr/include/python2.7/Python.h` is there and if not, get python-dev or whatever your distro uses as equivalent :)

Also, nuking "build" directory and running `python2 setup.py build` often helps with weird errors like that.

**Comment by LinuxJoeUser** [comment]
> You should be using master branch.

Whew!  Ok, good.  So I've been good about that.

> fatal error: Python.h: No such file or directory

No worries now, I'm not using Testing.

>  it shouldn't be on 3.16.xx

I'm not sure why it didn't update, but I went in and manually deleted all the scc/sc-controller stuff in /usr and re-git-cloned nd re-built and re-installed setup.py and I've got v4.0.1 installed now.

And now "untouched" etc. are available!  Yay!  And by setting LPAD press to `press(button(Keys.KEY_LEFTSHIFT` 

and Lpad to 

    dpad(
        button(Keys.KEY_W), 
        button(Keys.KEY_S),
        button(Keys.KEY_A),
        button(Keys.KEY_D)
    ) 
    and untouched(release(button(Keys.KEY_LEFTSHIFT)))
So all this is groovy!  :+1:
